### DISCUSSION TITLE
No unique identifiers in MQTT

### DISCUSSION THREAD

**Discussion by MikeBishop** [teslamate-org/teslamate#1568] [Ideas]
**Describe the bug**

MQTT data uses Teslamate-local IDs, but doesn't appear to expose a map of these IDs to specific vehicles.  The display name could perhaps be used for this purpose, but that's not guaranteed to be unique.

**Expected behavior**

If Teslamate published the vehicle's (owner API) ID or VIN as a persistent value, a listener can create a map of IDs to specific vehicles reliably.

**How to reproduce it (as minimally and precisely as possible)**:

Start listening to MQTT; try to figure out which vehicle is which without external context.

**Data**

Client mosq-hP1HWrVa3eVWVaf2uB sending CONNECT
Client mosq-hP1HWrVa3eVWVaf2uB received CONNACK (0)
Client mosq-hP1HWrVa3eVWVaf2uB sending SUBSCRIBE (Mid: 1, Topic: teslamate/#, QoS: 0, Options: 0x00)
Client mosq-hP1HWrVa3eVWVaf2uB received SUBACK
Subscribed (mid: 1): 0
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/geofence', ... (4 bytes))
Home
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/inside_temp', ... (4 bytes))
19.7
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/state', ... (6 bytes))
online
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/since', ... (27 bytes))
2021-04-19T16:17:22.740000Z
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/ideal_battery_range_km', ... (6 bytes))
384.02
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/sentry_mode', ... (5 bytes))
false
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/spoiler_type', ... (4 bytes))
None
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/heading', ... (2 bytes))
71
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/charge_port_door_open', ... (4 bytes))
true
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/charge_limit_soc', ... (2 bytes))
80
Client mosq-hP1HWrVa3eVWVaf2uB received PUBLISH (d0, q0, r1, m0, 'teslamate/cars/1/plugged_in', ... (4 bytes))
true

**Environment**

- TeslaMate version: 1.22.0
- Type of installation: Docker
- OS TeslaMate is installed on: Synology
- User OS & Browser: mosquitto_sub on Linux

**Comment by adriankumpf** [comment]
I consider this more of a feature request :) I'll move it to [Ideas 💡](https://github.com/adriankumpf/teslamate/discussions/categories/ideas).

**Comment by adriankumpf** [comment]
Two things come to my mind: A possible implementation should ensure that the current topics continue to exist, to not break existing installations. The vehicle and owner API IDs have changed in the past, so the VIN is likely the best of the three options.

**Comment by MikeBishop** [comment]
For right now, I'm using the logic that if Display Name matches and the odometer is within 1%, it's highly likely to be the same vehicle.  But I agree, VIN is likely to be the most permanent of the options.
